### ISSUE TITLE
tfd.Masked does not apply naturally to tfd.MarkovChain steps

### ISSUE THREAD

**Issue by xiaolong1979** [tensorflow/probability#1368]
Hi, I am trying to mask certain steps of a batch of Markov chains but cannot get it work with similar tfd.Masked call. Can some one show me how i should correct it? It would be helpful to include an example of masked MarkovChain in the documentation. Thanks a lot!

### generate a batche (2) of two dimensional random walks with tfd.MarkovChain, which works fine.
def transition_fn(_, previous_state):
    return tfd.JointDistributionNamedAutoBatched(
      batch_ndims=tf.rank(previous_state['X1']),
      model={
          'X1': tfd.Normal(loc=previous_state['X1'], scale=0.5),
          'X2': tfd.Normal(loc=previous_state['X2'], scale=1.0),
          })

process = tfd.MarkovChain(
    initial_state_prior=transition_fn(
        0, previous_state={
            'X1': [0.0,0.0],
            'X2': [0.0,0.0]}),
    transition_fn=transition_fn,
    num_steps=10)

print('Process batch shape',process.batch_shape)
print('Process event shape',process.event_shape)

x = process.sample()
lp = process.log_prob(x) 

print(x)

### create a mask to limit X1 and X2 less than or equal to 1.0
mask={'X2':tf.math.less(x['X2'], 1.0),
        'X1':tf.math.less(x['X1'], 1.0)}
print(mask)

### mask the process steps indicated by the mask 
process_masked = tfd.Masked(process,mask)  ### this is not working as expected

**Comment by brianwa84** [comment]
The Masked distribution's mask applies to batch dimensions, whereas it
sounds like you want to mask some parts of the *event* for a MarkovChain.
This would require custom support in MarkovChain similar to what is
provided in tfp.sts for masking observations.

**Comment by davmre** [comment]
I don't think `MarkovChain` itself will ever be able to support missing values directly. The distribution over the chain state at time `t` depends on the state at time `t - 1`; if any values are missing, the chain breaks.

The general way to handle missing values is to treat them as latent variables that you integrate over. In special cases like the models treated by `tfp.sts` and `tfd.LinearGaussianStateSpaceModel`, it can be possible to do that integration analytically using, e.g., linear Gaussian structure. Otherwise, you'd need to sample them using an approximate inference method like MCMC or VI. For example:


```python
def target_log_prob_fn(missing_xs):
  # Define a simple Gaussian chain. We could use `tfd.LinearGaussianStateSpaceModel`
  # to handle missing values directly, but the approach demonstrated here works for
  # more general chains.
  chain = tfd.MarkovChain(
    initial_state_prior=tfd.Normal(0., 1.),
    transition_fn=lambda _, x: tfd.Normal(x, 1.),
    num_steps=9)
  # Given the missing values we can evaluate the joint
  # density of the entire chain.
  return chain.log_prob(tf.concat([[1., 2., 3., 4.],
                                   missing_xs,
                                   [7., 8., 9.]], axis=0)

# Use MCMC to sample the missing values.
nuts_kernel = tfp.mcmc.NoUTurnSampler(target_log_prob_fn, ...)
sampled_missing_xs = tfp.mcmc.sample_chain(current_state=tf.zeros([2]), kernel=nuts_kernel, ...)
```

I'm going to close this as not really possible to fix in general, though feel free to reopen if you run into a more specific technical issue.

**Comment by xiaolong1979** [comment]
> The Masked distribution's mask applies to batch dimensions, whereas it sounds like you want to mask some parts of the *event* for a MarkovChain. This would require custom support in MarkovChain similar to what is provided in tfp.sts for masking observations.

Yes, you are right. I was thinking about truncating the event shape.

**Comment by xiaolong1979** [comment]
@davmre Thanks for your comments. I will check more on tfp.sts. 

It sounds like a customized log_prob method is required if valid event shapes are different among the batch. Is there a log_prob_parts equivalent for event components in MarkovChain? I wonder if we can set some of them as zero according to to event/step mask and then re-sum for log_prob.

I am more interesting in truncating the train, which is more specific than general masking. If we have log_prob_parts like other distributions, we may easily accommodate the truncated train.

**Comment by davmre** [comment]
@xiaolong1979 maybe the private `_log_prob_parts` method would do what you need? https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/distributions/markov_chain.py#L355

That's a good point that truncation should be feasible, even if general masking is not. Conceivably `MarkovChain` could accept an array of values for `num_steps`, and produce/accept sampled Tensors of step dimension `reduce_max(num_steps)` while masking dimensions with fewer steps. I doubt I'll have a chance to implement this in the immediate future, but I'd be happy to review a PR.

**Comment by xiaolong1979** [comment]
@davmre Cool! Let me check into it. Indeed, it may be possible to break the whole batch into multiple matches with different num_steps and do as usual before adding them together?

I am not so savvy in coding. Will try to learn how to call _log_prob_parts, which sounds like what's needed. Thanks a lot!

**Comment by xiaolong1979** [comment]
@davmre I tried to write a log_prob function for truncated chains, based on _log_prob_parts, and it appears working. Could you see if it is robust in general? 

### truncated trains log_prob
def log_prob_truncated(chains,data_obs,truncation_mask):
    """Returns log-prob for truncated Markov chains."""
    # chain = an object of tfd.MarkovChain
    # data_obs = sample for the chains
    # truncation_mask is a boolen with shape (num_step-1,batch_shape), where
    # num_step is the maximum step of the chain, derived from the sample data_obs
    # batch_shape is the batch_shape of the chain derived from the sample data_obs
    max_steps=data_obs[list(data_obs)[0]].shape[-1]-1
    batch_size=data_obs[list(data_obs)[0]].shape[-2]
    sample_shape=data_obs[list(data_obs)[0]].shape[-3]
    Truncation_mask_sample=tf.repeat(tf.reshape(Truncation_mask,shape=(max_steps,1,batch_size)), repeats=sample_shape, axis=1)
    parts_prob=chains._log_prob_parts(data_obs)
    return parts_prob[0]+tf.reduce_sum(parts_prob[1]*tf.cast(Truncation_mask_sample,dtype=tf.float32),axis=0)


### example of using the function with a constructed truncation mask

### two dimensional random walk
def transition_fn(_, previous_state):
    return tfd.JointDistributionNamedAutoBatched(
      batch_ndims=tf.rank(previous_state['X1']),
      model={
          'X1': tfd.Normal(loc=previous_state['X1'], scale=0.5),
          'X2': tfd.Normal(loc=previous_state['X2'], scale=1.0),
          })

max_steps=10

process = tfd.MarkovChain(
    initial_state_prior=transition_fn(
        0, previous_state={
            'X1': [0.0,0.0],
            'X2': [0.0,0.0]}),
    transition_fn=transition_fn,
    num_steps=max_steps)

print('Process batch shape',process.batch_shape)
print('Process event shape',process.event_shape)

sample_shape=3

x = process.sample(sample_shape)
lp = process.log_prob(x) 

### generate a truncation mask for the given sample x

import random

max_steps=x[list(x)[0]].shape[-1]-1
batch_size=x[list(x)[0]].shape[-2]
sample_shape=x[list(x)[0]].shape[-3]

mindex=tf.math.cumsum(tf.ones(shape=(max_steps,batch_size),dtype=tf.float32),axis=0)
truncation=tf.reshape(tf.convert_to_tensor([random.randrange(1, max_steps) 
                                 for i in range(batch_size)],dtype=tf.float32),shape=(1,batch_size))
Truncation_mask=tf.math.less(mindex, truncation)

print('Truncation mask \n',Truncation_mask)

### print truncated log_prob and log_rob without truncation
print('log_prob for the truncated trains \n',log_prob_truncated(process,x,Truncation_mask))
print('log_prob for the full trains \n',process.log_prob(x))
